release-19.1: pgerror: avoid wrapping special roachpb errors

knz opened the pull request:
Backport 1/1 commits from #35937.
Backport 1/1 commits from #35944.

/cc @cockroachdb/release

---

Fixes #35883.
Fixes #35943

knz committed:
pgerror: avoid wrapping special roachpb errors

Prior to this patch, the following scenario was possible:

1. some SQL function A() calls client.Txn() to call B() on its behalf
   with a retry loop.
2. client.Txn() calls B()
3. B() runs KV operation, encounters some *roachpb.RetryError
4. B() calls `pgerror.Wrap()` on that retry error, turns it into pg
   error with code 40001
5. B() returns into client.Txn()
6. client.Txn() tests the type of the error expecting to find a
   roachpb error for retries. Because it's a pgerror now, the test
   fails and the retry error is not handled automatically. It flows
   out.
7. A() receives a retry error when it was not expecting any.

The error would incur spurious DDL failures, with the following
example:

```
--- FAIL: TestLogic/fakedist-opt/computed (2.66s)
    logic.go:2346:

        testdata/logic_test/computed:687:
        expected:
        division by zero

        got:
        pq: column "d" being dropped, try again later
```

(This particular test would fail after just a few attempts during a
stress run.)

This patch fixes this scenario by ensuring that retry and ambiguous
result errors flow through `pgerror.Wrapf` without being flattened
into a `pgerror.Error`, and instead remain wrapped as the original
cause via `errors.Wrapf`.

The patch is verified to work by stress testing the logic tests and
verifying that the various error symptoms do not occur any more.

Additionally, two missing calls to `errors.Cause()` were added to
ensure any enclosed roachpb error is visible to a type check.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35939)
<!-- Reviewable:end -->

tbg commented:
LGTM but wait until the other one has merged please.

On Tue, Mar 19, 2019 at 2:55 PM kena <notifications@github.com> wrote:

> @knz <https://github.com/knz> requested your review on: #35939
> <https://github.com/cockroachdb/cockroach/pull/35939> release-19.1:
> pgerror: avoid wrapping special roachpb errors.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/35939#event-2213666016>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135GwH3yDtLjZjxw3aNwqDjZDWdOmMks5vYOxXgaJpZM4b8MX6>
> .
>

knz commented:
> LGTM but wait until the other one has merged please.

Works for me. I assume you also want a green master build?

tbg commented:
No, just for it to have merged to master (there have been cases where we
merged the backport before the original, and then some change had to be
made to the original, etc...)

On Tue, Mar 19, 2019 at 3:17 PM kena <notifications@github.com> wrote:

> LGTM but wait until the other one has merged please.
>
> Works for me. I assume you also want a green master build?
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/35939#issuecomment-474390219>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135M-l4tzJwoTQIWB0ibd8BsmG8gxSks5vYPF1gaJpZM4b8MX6>
> .
>

knz commented:
> No, just for it to have merged to master (there have been cases where we merged the backport before the original, and then some change had to be made to the original, etc...)

Oh I see. Well it's merged now.

knz commented:
... and we're getting another totally unrelated failure here. I'm going to check whether my patch is responsible for it.

knz committed:
distsqlpb,importccl: fix a regression in IMPORT

This patch ensures that the full error message (pre-unwrap) is
included for errors that flow out of distsql process towards the
gateway.

This is important because some other internal mechanisms inside
CockroachDB are expecting to see the prefixes introduced by
errors.Wrap to decide other things (a bad idea, but not fixable in the
short term), for example see issue #35942.

Release note: None

knz commented:
The new failure is #35943.
I fixed it in #35944, and added that commit to this backport PR.

I tested locally that stress testing the import code does not yield the error any more.

RaduBerinde commented:
LGTM
